### PR TITLE
Feature: Normalize

### PR DESCRIPTION
**Pull request by JakobEliasWagner**
# Feature: Normalize

## Description

This pull request introduces a new feature for data preprocessing: Z-Normalization. Z-Normalization, also known as standardization, is a technique used to scale data so that it has a mean of 0 and a standard deviation of 1. This process is essential in many machine learning algorithms as it ensures that each feature contributes equally to the analysis, improving the performance and stability of the models.

### Which issue does this PR tackle?

  - Addresses the need for a standardized data preprocessing step in our machine learning pipeline. The lack of a z-normalization feature has been a significant gap in our data preprocessing capabilities, leading to suboptimal performance in models sensitive to the scale of data.

### How does it solve the problem?

  - Implements Z-normalization.
  - Offers alias `Standardization`.
  - Offering compatibility with multi-dimensional tensors.

### How are the changes tested?

- Added unit tests:
  - 4 tests against trivial examples.
  - 1 test with common 8 students set.
  - 1 test testing for singularity.
  - 1 test assuring the alias is set correctly
  - 2 tests with higher dimensional data and multiple scaling dimensions

## Checklist for Contributors

- [x] Scope: This PR tackles exactly one problem.
- [x] Conventions: The branch follows the `feature/title-slug` convention.
- [x] Conventions: The PR title follows the `Bugfix: Title` convention.
- [x] Coding style: The code passes all pre-commit hooks.
- [x] Documentation: All changes are well-documented.
- [x] Tests: New features are tested and all tests pass successfully.
- [x] Changelog: Updated CHANGELOG.md for new features or breaking changes.
- [x] Review: A suitable reviewer has been assigned.

## Checklist for Reviewers:

- [x] The PR solves the issue it claims to solve and only this one.
- [x] Changes are tested sufficiently and all tests pass.
- [x] Documentation is complete and well-written.
- [x] Changelog has been updated, if necessary.

### CODE REVIEW

File: src/continuity/transforms/__init__.py [hunk 1]
```diff
@@ -1,7 +1,13 @@
 from .transform import Transform
 from .compose import Compose
+from .scaling import ZNormalization
+
+# aliases
+Standardization = ZNormalization
```
[the remark this answers is not in the document]
Comment (JakobEliasWagner) [reply src/continuity/transforms/__init__.py]: Renamed it to Normalize

File: src/continuity/transforms/scaling.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+
+class Normalize(Transform):
+    r"""Z-normalization transformation.
+
+    This transformation uses the mean $\mu$ and the standard deviation $\sigma$ passed in the initialization and scales
+    tensors $x$ with the mapping $z(x)=\frac{x - \mu}{\sigma}$.
+
+    Attributes:
+        mean: mean value of the tensor.
```
Comment (samuelburbulla) [inline src/continuity/transforms/scaling.py:14]: Remove "of the tensor"

File: src/continuity/transforms/scaling.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        std: standard deviation of the tensor.
```
Comment (samuelburbulla) [inline src/continuity/transforms/scaling.py:15]: Remove "of the tensor"

File: src/continuity/transforms/scaling.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        mean: torch.Tensor,
+        std: torch.Tensor,
+        epsilon: float = None,
+    ):
+        """
+
+        Args:
+            mean: mean used to scale tensors in forward. To scale a (100, 40, 3) tensor along the last
+            dimension, provide the mean along the last dimension in the shape (1, 1, 3).
+            std: standard deviation used to scale tensors in forward. Should have the same shape as the mean.
```
Comment (samuelburbulla) [inline src/continuity/transforms/scaling.py:30]: Hm, docstring says "Should have the same shape", but shape is checked (=enforced) by the assertion in line 33. I think we should let the freedom to have different shapes, e.g., a channel-wise mean but a scalar std or so. Remove assertion and comment in docstring.

File: src/continuity/transforms/scaling.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            epsilon: Value for numerical stability (to prevent divide by zero).
+        """
+        assert mean.shape == std.shape
+        assert torch.all(torch.greater_equal(std, torch.zeros(std.shape)))
+        super().__init__()
+        self.mean = mean
+        if torch.allclose(std, torch.zeros(std.shape)):
```
Comment (samuelburbulla) [inline src/continuity/transforms/scaling.py:37]: We already introduced eps to prevent division by zero, why then throw a warning? It's similar to what we discussed in the transform, I assume the users nows what he is doing here, we shouldn't have to take care of misuse. Remove warning.

File: src/continuity/transforms/scaling.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+        Args:
+            tensor: input.
+
+        Returns:
+            Z-normalized tensor.
+        """
+        return (tensor - self.mean) / (self.std + self.epsilon)
+
+    def backward(self, tensor: torch.Tensor) -> torch.Tensor:
```
Comment (samuelburbulla) [inline src/continuity/transforms/scaling.py:61]: docstring?

**Review by samuelburbulla — CHANGES_REQUESTED** [review]
Better, but I still found some minor issues.

File: src/continuity/transforms/scaling.py (hunk 6 above)
Comment (JakobEliasWagner) [reply src/continuity/transforms/scaling.py]: Plus the method name followed the old structure.

File: src/continuity/transforms/scaling.py (hunk 5 above)
Comment (JakobEliasWagner) [reply src/continuity/transforms/scaling.py:35]: Numerical stability with eps is only ensured when all values of the standard deviation are actually bigger than or equal to zero. This may lead to division by zero.

Comment (JakobEliasWagner) [reply src/continuity/transforms/scaling.py:35]: Thus my suggestion would be to adjust the method in the following way:

- Remove eps entirely
- Replace zero values in std with eps

This would enable negative std and ensure no `inf` value is introduced as an input.

**Comment by JakobEliasWagner** [comment]
The reason as to why this test fails is unrelated to eps. Removed the comment in 
image

**Comment by JakobEliasWagner** [comment]
The previous plot is related to the range of the scaled tensor. The tests are performed for a range in the order of one -> 1e-7 is appropriate for testing.
image

**Comment by samuelburbulla** [comment]
I improved the docs a little and set the random seed, because tests failed from time to time.
